File: 2023/03/pg_connecting_the_islands.py
def solution(n, costs):
    def find_set(x):
        if p[x] == x:
            return x
        else:
            p[x] = find_set(p[x])
            return p[x]

    def union(x, y):
        parent_x = find_set(x)
        parent_y = find_set(y)
        if parent_x > parent_y:
            p[parent_y] = parent_x
        else:
            p[parent_x] = parent_y

    costs.sort(key=lambda x: x[2])

    # make_set: 모든 정점에 대해 집합 생성
    p = [i for i in range(n)]

    cnt = 0
    result = 0
    mst = []
    # 모든 간선에 대해서 반복 -> V-1개의 간선이 선택될 때까지

    for i in range(len(costs)):
        s, e, c = costs[i]
        print(s, e)
        # 사이클이면 스킵: 간선의 두 정점이 서로 같은 집합이면 => find_set

        if find_set(s) == find_set(e):
            continue
        # 간선 선택
        # => mst에 간선 정보 더하기 / 두 정점을 합친다 => union
        result += c
        union(s, e)
        print("union p: ", p)
        # 연결한 간선 정보를 mst 리스트에 담아보자.
        mst.append(costs[i])
        cnt += 1

        if cnt == n - 1:
            break

    return result

n = 4
costs = [[0,1,1],[0,2,2],[1,2,5],[1,3,1],[2,3,8]]

solution(n, costs)

File: 2023/03/pg_connecting_the_islands.js
const solution = (n, costs) => {
  const findSet = (x) => {
    if (p[x] === x) {
      return x;
    } else {
      p[x] = findSet(p[x]);
      return p[x];
    }
  };

  const union = (x, y) => {
    const px = findSet(x);
    const py = findSet(y);

    if (rank[px] > rank[py]) {
      p[py] = px;
    } else {
      p[px] = py;
      if (rank[px] === rank[py]) {
        rank[py] += 1;
      }
    }
  };

  // 정점의 개수, 관례적으로 사용하는 V로 재정의해준다.
  const V = n;
  //
  const E = costs.length;

  // 간선의 가중치를 기준으로 정렬한다.
  costs.sort((a, b) => {
    return a[2] > b[2] ? 1 : -1;
  });

  // 배열 상태를 확인
  console.log(costs);

  const p = new Array(E).fill(0).map((elm, idx) => idx);
  const rank = new Array(E).fill(0);

  let cnt = 0;
  let result = 0;

  for (let i = 0; i < E; i += 1) {
    const [s, e, c] = costs[i];

    if (findSet(s) === findSet(e)) {
      continue;
    }

    result += c;
    union(s, e);

    cnt += 1;
    if (cnt === V - 1) {
      break;
    }
  }

  return result;
};

// Test Cases
const n = 4;
const costs = [
  [0, 1, 1],
  [0, 2, 2],
  [1, 2, 5],
  [1, 3, 1],
  [2, 3, 8],
];
const result = 4;

console.log(solution(n, costs) === result);
